File: backend/api/views/reports.py
from django.db.models import Count, Q
from django.db.models.functions import TruncMonth
from rest_framework.decorators import api_view, permission_classes
from rest_framework.permissions import IsAuthenticated
from rest_framework.response import Response

from api.models import AuditLog, Employee, Training, User
from api.serializers import AuditLogSerializer
from api.utils import days_from_today, today, user_display_name


@api_view(['GET'])
@permission_classes([IsAuthenticated])
def overview(request):
    current = today()
    in10 = days_from_today(10)

    return Response({
        'totalEmployees': Employee.objects.filter(status='active').count(),
        'totalTrainings': Training.objects.count(),
        'expiredCerts': Training.objects.filter(
            expiration_date__isnull=False,
            expiration_date__lt=current,
        ).count(),
        'expiring30': Training.objects.filter(
            expiration_date__isnull=False,
            expiration_date__gte=current,
            expiration_date__lte=in10,
        ).count(),
        'expiring60': Training.objects.filter(
            expiration_date__isnull=False,
            expiration_date__gte=current,
            expiration_date__lte=in10,
        ).count(),
        'totalUsers': User.objects.filter(status='active').count(),
    })


@api_view(['GET'])
@permission_classes([IsAuthenticated])
def by_category(request):
    rows = (
        Training.objects.exclude(category='')
        .values('category')
        .annotate(count=Count('id'))
        .order_by('-count')
    )
    return Response(list(rows))


@api_view(['GET'])
@permission_classes([IsAuthenticated])
def by_factory(request):
    rows = []
    factories = (
        Employee.objects.filter(status='active')
        .exclude(factory='')
        .values_list('factory', flat=True)
        .distinct()
        .order_by('factory')
    )
    for factory in factories:
        employee_count = Employee.objects.filter(status='active', factory=factory).count()
        training_count = Training.objects.filter(employee__factory=factory).count()
        rows.append({
            'factory': factory,
            'employee_count': employee_count,
            'training_count': training_count,
        })
    return Response(rows)


@api_view(['GET'])
@permission_classes([IsAuthenticated])
def expiring(request):
    current = today()
    in10 = days_from_today(10)
    days = request.query_params.get('days')
    expired_only = request.query_params.get('expired') == 'true'

    if expired_only:
        trainings = (
            Training.objects.select_related('employee')
            .filter(expiration_date__isnull=False, expiration_date__lt=current)
            .order_by('expiration_date')
        )
    elif days:
        in_days = days_from_today(int(days))
        trainings = (
            Training.objects.select_related('employee')
            .filter(
                expiration_date__isnull=False,
                expiration_date__gte=current,
                expiration_date__lte=in_days,
            )
            .order_by('expiration_date')
        )
    else:
        trainings = (
            Training.objects.select_related('employee')
            .filter(expiration_date__isnull=False, expiration_date__lte=in10)
            .order_by('expiration_date')[:100]
        )

    rows = []
    for training in trainings:
        if training.expiration_date < current:
            cert_status = 'expired'
        elif training.expiration_date <= in10:
            cert_status = 'expiring'
        else:
            cert_status = 'valid'

        rows.append({
            'id': training.id,
            'title': training.title,
            'expiration_date': training.expiration_date.isoformat(),
            'training_date': training.training_date.isoformat(),
            'full_name': training.employee.full_name,
            'emp_code': training.employee.employee_id,
            'factory': training.employee.factory,
            'line': training.employee.line,
            'team': training.employee.team,
            'cert_status': cert_status,
        })

    return Response(rows)


@api_view(['GET'])
@permission_classes([IsAuthenticated])
def audit_logs(request):
    limit = int(request.query_params.get('limit') or 50)
    logs = AuditLog.objects.select_related('user').order_by('-created_at')[:limit]
    return Response(AuditLogSerializer(logs, many=True).data)


@api_view(['GET'])
@permission_classes([IsAuthenticated])
def record_audit_logs(request, table_name, record_id):
    import json as _json
    logs = (
        AuditLog.objects
        .select_related('user')
        .filter(table_name=table_name, record_id=record_id)
        .order_by('-created_at')
    )
    result = []
    for log in logs:
        entry = {
            'action': log.action,
            'full_name': user_display_name(log.user, log.username),
            'created_at': log.created_at.strftime('%b %d, %Y %I:%M %p'),
            'summary': log.details or '',
            'changes': {},
        }
        # Try parsing JSON details for before/after
        if log.details:
            try:
                parsed = _json.loads(log.details)
                entry['summary'] = parsed.get('summary', log.details)
                entry['changes'] = parsed.get('changes', {})
            except (ValueError, TypeError):
                entry['summary'] = log.details
        result.append(entry)
    return Response(result)


@api_view(['GET'])
@permission_classes([IsAuthenticated])
def export_trainings(request):
    qs = Training.objects.select_related('employee').all()

    category = request.query_params.get('category')
    factory = request.query_params.get('factory')
    search = request.query_params.get('search')
    worker_line_status = request.query_params.get('worker_line_status')
    take = request.query_params.get('take')
    training_date = request.query_params.get('training_date')
    emp_status = request.query_params.get('status')

    if factory:
        qs = qs.filter(employee__factory=factory)
    if category:
        qs = qs.filter(category=category)
    if worker_line_status:
        qs = qs.filter(worker_line_status=worker_line_status)
    if take:
        qs = qs.filter(take=int(take))
    if training_date:
        qs = qs.filter(training_date=training_date)
    date_from = request.query_params.get('date_from')
    date_to = request.query_params.get('date_to')
    if date_from:
        qs = qs.filter(training_date__gte=date_from)
    if date_to:
        qs = qs.filter(training_date__lte=date_to)
    if emp_status:
        qs = qs.filter(employee__status=emp_status)
    if search:
        qs = qs.filter(
            Q(employee__full_name__icontains=search)
            | Q(employee__employee_id__icontains=search)
            | Q(title__icontains=search)
        )

    current = today()
    in10 = days_from_today(10)
    if request.query_params.get('expired') == 'true':
        qs = qs.filter(expiration_date__isnull=False, expiration_date__lt=current)
    elif request.query_params.get('expiring_soon') == 'true':
        qs = qs.filter(
            expiration_date__isnull=False,
            expiration_date__gte=current,
            expiration_date__lte=in10,
        )

    rows = []
    for training in qs.order_by('employee__full_name', '-training_date'):
        employee = training.employee
        rows.append({
            'id': training.id,
            'emp_code': employee.employee_id,
            'full_name': employee.full_name,
            'factory': employee.factory,
            'line': employee.line,
            'team': employee.team,
            'title': training.title,
            'category': training.category,
            'training_date': training.training_date.isoformat(),
            'trainer': training.trainer,
            'validity_months': training.validity_months,
            'expiration_date': (
                training.expiration_date.isoformat() if training.expiration_date else None
            ),
            'process_classification': training.process_classification,
            'worker_line_status': training.worker_line_status,
            'take': training.take,
            'remarks': training.remarks,
        })

    return Response(rows)


@api_view(['GET'])
@permission_classes([IsAuthenticated])
def takes_per_month(request):
    """
    Returns data grouped by month (Y) and take number (X).
    Each entry: { month, takes: [{take, count}, ...] }
    """
    MONTH_NAMES = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
                   'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']

    rows = (
        Training.objects
        .annotate(month=TruncMonth('training_date'))
        .values('month', 'take')
        .annotate(count=Count('id'))
        .order_by('month', 'take')
    )

    # Build month → {take: count} map
    from collections import defaultdict
    month_map = defaultdict(dict)
    month_order = []
    for row in rows:
        if not row['month']:
            continue
        label = MONTH_NAMES[row['month'].month - 1] + ' ' + str(row['month'].year)
        if label not in month_map:
            month_order.append(label)
        month_map[label][row['take']] = row['count']

    # All take numbers that appear
    all_takes = sorted({t for m in month_map.values() for t in m})

    return Response({
        'months': month_order,
        'takes': all_takes,
        'data': {
            month: {str(t): month_map[month].get(t, 0) for t in all_takes}
            for month in month_order
        },
    })

File: client/electron/db/reports.js
'use strict';

const { getPool } = require('./pool');
const { requireAuth } = require('./auth');
const { today, daysFromToday, userDisplayName, formatAuditDate, MONTH_NAMES_SHORT, normalizeCertRecert } = require('./helpers');

async function scalar(pool, sql, vals = []) {
  const [rows] = await pool.query(sql, vals);
  return Number(rows[0].c);
}

async function overview() {
  requireAuth();
  const pool = getPool();
  const current = today();
  const in10 = daysFromToday(10);

  const totalEmployees = await scalar(pool, "SELECT COUNT(*) c FROM employees WHERE status = 'active'");
  const employeeTotal = await scalar(pool, 'SELECT COUNT(*) c FROM employees');
  const totalTrainings = await scalar(pool, 'SELECT COUNT(*) c FROM trainings WHERE is_archived = 0');
  const activeCertifications = await scalar(pool,
    'SELECT COUNT(*) c FROM trainings WHERE is_archived = 0 AND (expiration_date IS NULL OR expiration_date >= ?)', [current]);
  const expiredCerts = await scalar(pool,
    'SELECT COUNT(*) c FROM trainings WHERE is_archived = 0 AND expiration_date IS NOT NULL AND expiration_date < ?', [current]);
  const expiringCount = await scalar(pool,
    'SELECT COUNT(*) c FROM trainings WHERE is_archived = 0 AND expiration_date IS NOT NULL AND expiration_date >= ? AND expiration_date <= ?', [current, in10]);
  const employeesWithTraining = await scalar(pool,
    "SELECT COUNT(DISTINCT e.id) c FROM employees e JOIN trainings t ON t.employee_id = e.id AND t.is_archived = 0 WHERE e.status = 'active'");
  const totalUsers = await scalar(pool, "SELECT COUNT(*) c FROM users WHERE status = 'active'");

  const completionRate = totalEmployees
    ? Math.round((employeesWithTraining / totalEmployees) * 100 * 10) / 10
    : 0.0;

  return {
    totalEmployees,
    employeeTotal,
    totalTrainings,
    activeCertifications,
    trainingCompletionRate: completionRate,
    expiredCerts,
    expiring30: expiringCount,
    expiredAndExpiring: expiredCerts + expiringCount,
    expiring60: expiringCount,
    totalUsers,
  };
}

async function byCategory() {
  requireAuth();
  const pool = getPool();
  const [rows] = await pool.query(
    "SELECT category, COUNT(*) AS count FROM trainings WHERE category <> '' GROUP BY category ORDER BY count DESC"
  );
  return rows.map((r) => ({ category: r.category, count: Number(r.count) }));
}

async function byFactory() {
  requireAuth();
  const pool = getPool();
  const [factories] = await pool.query(
    "SELECT DISTINCT factory FROM employees WHERE status = 'active' AND factory <> '' ORDER BY factory"
  );
  const result = [];
  for (const { factory } of factories) {
    const employeeCount = await scalar(pool, "SELECT COUNT(*) c FROM employees WHERE status = 'active' AND factory = ?", [factory]);
    const trainingCount = await scalar(pool,
      'SELECT COUNT(*) c FROM trainings t JOIN employees e ON e.id = t.employee_id WHERE e.factory = ?', [factory]);
    result.push({ factory, employee_count: employeeCount, training_count: trainingCount });
  }
  return result;
}

async function expiring(params = {}) {
  requireAuth();
  const pool = getPool();
  const current = today();
  const in10 = daysFromToday(10);

  const base = `SELECT t.id, t.title, t.expiration_date, t.training_date,
    e.full_name, e.employee_id AS emp_code, e.factory, e.line, e.team
    FROM trainings t JOIN employees e ON e.id = t.employee_id`;

  let sql;
  let vals = [];
  if (params.expired === 'true') {
    sql = `${base} WHERE t.expiration_date IS NOT NULL AND t.expiration_date < ? ORDER BY t.expiration_date`;
    vals = [current];
  } else if (params.days) {
    const inDays = daysFromToday(parseInt(params.days, 10));
    sql = `${base} WHERE t.expiration_date IS NOT NULL AND t.expiration_date >= ? AND t.expiration_date <= ? ORDER BY t.expiration_date`;
    vals = [current, inDays];
  } else {
    sql = `${base} WHERE t.expiration_date IS NOT NULL AND t.expiration_date <= ? ORDER BY t.expiration_date LIMIT 100`;
    vals = [in10];
  }

  const [rows] = await pool.query(sql, vals);
  return rows.map((t) => {
    const exp = String(t.expiration_date).slice(0, 10);
    let certStatus;
    if (exp < current) certStatus = 'expired';
    else if (exp <= in10) certStatus = 'expiring';
    else certStatus = 'valid';
    return {
      id: t.id,
      title: t.title,
      expiration_date: exp,
      training_date: String(t.training_date).slice(0, 10),
      full_name: t.full_name,
      emp_code: t.emp_code,
      factory: t.factory,
      line: t.line,
      team: t.team,
      cert_status: certStatus,
    };
  });
}

async function auditLogs({ limit } = {}) {
  requireAuth();
  const pool = getPool();
  const lim = parseInt(limit, 10) || 50;
  const [rows] = await pool.query(
    `SELECT a.id, a.action, a.table_name, a.record_id, a.details, a.created_at, a.username,
            u.full_name AS user_full_name, u.username AS user_username
     FROM audit_logs a LEFT JOIN users u ON u.id = a.user_id
     ORDER BY a.created_at DESC LIMIT ?`,
    [lim]
  );
  return rows.map((r) => ({
    id: r.id,
    action: r.action,
    table_name: r.table_name,
    record_id: r.record_id,
    details: r.details,
    created_at: r.created_at ? String(r.created_at).replace(' ', 'T') : null,
    full_name: userDisplayName(
      r.user_username ? { full_name: r.user_full_name, username: r.user_username } : null,
      r.username
    ),
  }));
}

async function recordLogs({ table, id } = {}) {
  requireAuth();
  const pool = getPool();
  const [rows] = await pool.query(
    `SELECT a.action, a.details, a.created_at, a.username,
            u.full_name AS user_full_name, u.username AS user_username
     FROM audit_logs a LEFT JOIN users u ON u.id = a.user_id
     WHERE a.table_name = ? AND a.record_id = ?
     ORDER BY a.created_at DESC`,
    [table, id]
  );
  return rows.map((log) => {
    const entry = {
      action: log.action,
      full_name: userDisplayName(
        log.user_username ? { full_name: log.user_full_name, username: log.user_username } : null,
        log.username
      ),
      created_at: formatAuditDate(log.created_at),
      summary: log.details || '',
      changes: {},
    };
    if (log.details) {
      try {
        const parsed = JSON.parse(log.details);
        if (parsed && typeof parsed === 'object') {
          entry.summary = parsed.summary != null ? parsed.summary : log.details;
          entry.changes = parsed.changes || {};
        }
      } catch (_) {
        entry.summary = log.details;
      }
    }
    return entry;
  });
}

async function exportTrainings(params = {}) {
  requireAuth();
  const { ensureSchemaReady } = require('./pool');
  await ensureSchemaReady();
  const pool = getPool();
  const where = ['t.is_archived = 0'];
  const vals = [];

  if (params.factory) { where.push('e.factory = ?'); vals.push(params.factory); }
  if (params.category) { where.push('t.category = ?'); vals.push(params.category); }
  if (params.title) { where.push('t.title LIKE ?'); vals.push(`%${params.title}%`); }
  if (params.worker_line_status) { where.push('t.worker_line_status = ?'); vals.push(params.worker_line_status); }
  if (params.cert_recert) { where.push('t.cert_recert = ?'); vals.push(normalizeCertRecert(params.cert_recert)); }
  if (params.pass_fail) { where.push('t.pass_fail = ?'); vals.push(params.pass_fail === 'Failed' ? 'Failed' : 'Passed'); }
  if (params.take) { where.push('t.take = ?'); vals.push(parseInt(params.take, 10)); }
  if (params.training_date) { where.push('t.training_date = ?'); vals.push(params.training_date); }
  if (params.date_from) { where.push('t.training_date >= ?'); vals.push(params.date_from); }
  if (params.date_to) { where.push('t.training_date <= ?'); vals.push(params.date_to); }
  if (params.status) { where.push('e.status = ?'); vals.push(params.status); }
  if (params.search) {
    where.push('(e.full_name LIKE ? OR e.employee_id LIKE ? OR t.title LIKE ?)');
    const like = `%${params.search}%`;
    vals.push(like, like, like);
  }

  const current = today();
  const in10 = daysFromToday(10);
  if (params.expired === 'true') {
    where.push('t.expiration_date IS NOT NULL AND t.expiration_date < ?');
    vals.push(current);
  } else if (params.expiring_soon === 'true') {
    where.push('t.expiration_date IS NOT NULL AND t.expiration_date >= ? AND t.expiration_date <= ?');
    vals.push(current, in10);
  }

  const sql = `SELECT t.id, e.employee_id AS emp_code, e.full_name, e.factory, e.line, e.team,
    t.title, t.category, t.training_date, t.trainer, t.validity_months, t.expiration_date,
    t.process_classification, t.worker_line_status, t.cert_recert, t.pass_fail, t.take, t.remarks
    FROM trainings t JOIN employees e ON e.id = t.employee_id
    WHERE ${where.join(' AND ')} ORDER BY e.full_name ASC, t.training_date DESC`;
  const [rows] = await pool.query(sql, vals);
  return rows.map((t) => ({
    id: t.id,
    emp_code: t.emp_code,
    full_name: t.full_name,
    factory: t.factory,
    line: t.line,
    team: t.team,
    title: t.title,
    category: t.category,
    training_date: String(t.training_date).slice(0, 10),
    trainer: t.trainer,
    validity_months: t.validity_months,
    expiration_date: t.expiration_date ? String(t.expiration_date).slice(0, 10) : null,
    process_classification: t.process_classification,
    worker_line_status: t.worker_line_status,
    cert_recert: normalizeCertRecert(t.cert_recert),
    pass_fail: t.pass_fail === 'Failed' ? 'Failed' : 'Passed',
    take: t.take,
    remarks: t.remarks,
  }));
}

async function takesPerMonth() {
  requireAuth();
  const pool = getPool();
  const [rows] = await pool.query(
    `SELECT YEAR(training_date) AS year, MONTH(training_date) AS month_num, take, COUNT(*) AS count
     FROM trainings
     WHERE is_archived = 0 AND take IN (1, 2, 3)
     GROUP BY year, month_num, take
     ORDER BY year, month_num, take`
  );

  const allTakes = [1, 2, 3];
  const monthMap = {};
  const monthOrder = [];
  for (const row of rows) {
    if (!row.year || !row.month_num) continue;
    const label = `${MONTH_NAMES_SHORT[row.month_num - 1]} ${row.year}`;
    if (!(label in monthMap)) {
      monthMap[label] = {};
      monthOrder.push(label);
    }
    monthMap[label][row.take] = Number(row.count);
  }

  const data = {};
  for (const month of monthOrder) {
    data[month] = {};
    for (const t of allTakes) {
      data[month][String(t)] = monthMap[month][t] || 0;
    }
  }

  return { months: monthOrder, takes: allTakes, data };
}

module.exports = {
  overview, byCategory, byFactory, expiring,
  auditLogs, recordLogs, exportTrainings, takesPerMonth,
};
